### PR TITLE
feat: [PL-39556]: Added api to enable public access on a resource in open api

### PR DESCRIPTION
**Pull request by meenakshi-harness**
## Describe your changes

As part of this PR we have added endpoint for enabling public access on a resource in ACS. 
We have marked API as internal for now.

## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51808)

### CODE REVIEW

**Comment by meenakshi-harness** [comment]
trigger uts

**Comment by meenakshi-harness** [comment]
trigger ut5

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51808)* status: 5 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`access-control/contracts/openapi/v1/openapi.yaml` line 722 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQOF6AFgGL-Ea5UvC:-NcLQOF6AFgGL-Ea5UvD:b-dteery) ([raw file](https://github.com/harness/harness-core/blob/24bd602ffc43f10de9df26b34e2d70e37a41f5cb/access-control/contracts/openapi/v1/openapi.yaml#L722)):*
> ```YAML
>           type: boolean
>           description: Is harness managed
>     ResourceScope:
> ```

Can this not be reused? This is a very generic concept.
___
*[`access-control/contracts/openapi/v1/openapi.yaml` line 740 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQU4n2lmCN5zExZDd:-NcLQU4n2lmCN5zExZDe:b-ia6wzh) ([raw file](https://github.com/harness/harness-core/blob/24bd602ffc43f10de9df26b34e2d70e37a41f5cb/access-control/contracts/openapi/v1/openapi.yaml#L740)):*
> ```YAML
>       required:
>         - accountIdentifier
>     PublicAccessRequest:
> ```

How will we disable public access to a resource?
___
*[`access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java` line 22 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQspq33lFgyZG86TL:-NcLQspq33lFgyZG86TM:b-zaajwn) ([raw file](https://github.com/harness/harness-core/blob/24bd602ffc43f10de9df26b34e2d70e37a41f5cb/access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java#L22)):*
> ```Java
> @Builder
> @EqualsAndHashCode
> public class PublicAccessAssignmentMapping {
> ```

PublicAccessRoleAssignmentMapping
___
*[`access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java` line 22 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQwU00U5rKwPaPC7Y:-NcLQwU00U5rKwPaPC7Z:b-1lutcp) ([raw file](https://github.com/harness/harness-core/blob/24bd602ffc43f10de9df26b34e2d70e37a41f5cb/access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java#L22)):*
> ```Java
> @Builder
> @EqualsAndHashCode
> public class PublicAccessAssignmentMapping {
> ```

This class wouldn't get persisted?
How would we make sure if resource is deleted, we delete it from Access Control as well?
___
*[`access-control/service/src/main/resources/io/harness/accesscontrol/resources/public-access-assignment-mapping.txt` line 1 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQnZp2vmuC9aC4QzR:-NcLQnZp2vmuC9aC4QzS:b2ze2mx) ([raw file](https://github.com/harness/harness-core/blob/24bd602ffc43f10de9df26b34e2d70e37a41f5cb/access-control/service/src/main/resources/io/harness/accesscontrol/resources/public-access-assignment-mapping.txt#L1)):*
> ```help
> publicAccessAssignmentMapping:
> ```

publicAccessRoleAssignmentMapping

**Review by meenakshi-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51808)* status: 5 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`access-control/contracts/openapi/v1/openapi.yaml` line 722 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQOF6AFgGL-Ea5UvC:-NcMuTdoB2yxe541c4i1:b-7zhdx) ([raw file](https://github.com/harness/harness-core/blob/50e182c7acbc39456da013b72374beed46088c1e/access-control/contracts/openapi/v1/openapi.yaml#L722)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Can this not be reused? This is a very generic concept.
</blockquote></details>

Yes it can be re-used. But this is first time we are defining it in ACS open api yaml. Let me re-name it to Scope to make it more generic.
___
*[`access-control/contracts/openapi/v1/openapi.yaml` line 740 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQU4n2lmCN5zExZDd:-NcMudyj7PucUjM90iHG:b-abtx8l) ([raw file](https://github.com/harness/harness-core/blob/50e182c7acbc39456da013b72374beed46088c1e/access-control/contracts/openapi/v1/openapi.yaml#L740)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

How will we disable public access to a resource?
</blockquote></details>

We will take disable API separately. This is only to enable public access
___
*[`access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java` line 22 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQspq33lFgyZG86TL:-NcMulPN4b76u5C4ppf6:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/50e182c7acbc39456da013b72374beed46088c1e/access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java#L22)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

PublicAccessRoleAssignmentMapping
</blockquote></details>

Done.
___
*[`access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java` line 22 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQwU00U5rKwPaPC7Y:-NcMuxcYEYAm6wXy0MMx:bhzhyhq) ([raw file](https://github.com/harness/harness-core/blob/50e182c7acbc39456da013b72374beed46088c1e/access-control/libs/core/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessAssignmentMapping.java#L22)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

This class wouldn't get persisted?
How would we make sure if resource is deleted, we delete it from Access Control as well?
</blockquote></details>

Didn't get this. This is only to keep mapping of what resourceType we are allowing for public access. This should have affect on resource getting deleted/created.
___
*[`access-control/service/src/main/resources/io/harness/accesscontrol/resources/public-access-assignment-mapping.txt` line 1 at r1](https://reviewable.io/reviews/harness/harness-core/51808#-NcLQnZp2vmuC9aC4QzR:-NcMvE9j8g2RTCCV8dbQ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/50e182c7acbc39456da013b72374beed46088c1e/access-control/service/src/main/resources/io/harness/accesscontrol/resources/public-access-assignment-mapping.txt#L1)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

publicAccessRoleAssignmentMapping
</blockquote></details>

Done.

**Comment by meenakshi-harness** [comment]
trigger uts

**Comment by meenakshi-harness** [comment]
trigger uts

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 5 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51808#-:-NcROZKkDzvuXDyC8yS3:bnfp4nl)* status: 5 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by meenakshi-harness** [comment]
trigger messagecheck

**Comment by meenakshi-harness** [comment]
trigger ut1

**Review by piyush-harness — APPROVED** [review]
> Reviewed 1 of 5 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51808)* status: 1. unresolved discussions.

File: access-control/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -701,6 +719,43 @@ components:
...
           type: boolean
           description: Is harness managed
+    PublicAccessRequest:
+      title: PublicAccessRequest
+      x-stoplight:
+        id: iwewenxaom066
+      type: object
+      description: Request to mark a resource as public
+      properties:
+        resourceIdentifier:
```
Comment (mankrit-singh) [inline access-control/contracts/openapi/v1/openapi.yaml:729]: please change the casing to snake case eg. resource_identifier
Same for other parameters

File: access-control/contracts/openapi/v1/openapi.yaml [hunk 2]
```diff
@@ -701,6 +719,43 @@ components:
...
+          description: Account Identifier
+        orgIdentifier:
+          type: string
+          description: Org Identifier
+        projectIdentifier:
+          type: string
+          description: Project Identifier
+      required:
+        - accountIdentifier
+      description: Scope of the Ressource
```
Comment (mankrit-singh) [inline access-control/contracts/openapi/v1/openapi.yaml:758]: typo on resource

**Comment by meenakshi-harness** [comment]
trigger uts

**Comment by meenakshi-harness** [comment]
trigger ut5

**Comment by meenakshi-harness** [comment]
trigger ut7

**Comment by meenakshi-harness** [comment]
trigger ut5

**Review by Jimit09Gandhi — APPROVED** [review]
> Reviewed 1 of 5 files at r1, 3 of 3 files at r2, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51808)* status: 1 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by ashishsanodia-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 5 files at r1, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51808#-:-Nd-z1pzAJt_ZCKp5-eB:bnfp4nl)* status: 1. unresolved discussions.

**Pull request merged**